% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/LineagePlot.R
\name{PlotBayesFactorHeatmap}
\alias{PlotBayesFactorHeatmap}
\title{Plot Heatmap of Bayes Factor for Variants of Interest}
\usage{
PlotBayesFactorHeatmap(
  voi,
  N,
  af.dm,
  spatial_coords,
  BF_cap = 3,
  cor_order = T,
  ngroups = length(voi)
)
}
\arguments{
\item{voi}{Character vector, list of variants of interest.}

\item{N}{Matrix (voi x spot), total counts per variant per spot.}

\item{af.dm}{Matrix (voi x spot), allele frequency data per variant per spot.}

\item{spatial_coords}{Data frame (spot x 2), spatial coordinates (SPATIAL_1, SPATIAL_2).}

\item{BF_cap}{Numeric, maximum value to cap the Bayes Factor (default: 3).}

\item{ngroups}{Numeric, number of groups to cut the variant dendrogram (default: number of variants).}
}
\value{
A heatmap showing the Bayes Factor for each variant.
}
\description{
This function computes and visualizes the Bayes Factor (BF) for a list of variants
in a spatial transcriptomics dataset. The BF is calculated using a binomial likelihood
ratio test, comparing a baseline probability (\code{p0}) estimated from all cells to the
maximum likelihood estimate (\code{p1}) for each spatial spot.
}
\details{
To order the variants by selection process as MAESTER:
variants are first ordered by hierarchical clustering and
the spots are iteratively reâ€ordered such that those with higher BF in each variant group
are placed together.
}
\examples{
PlotBayesFactorHeatmap(c("3054_G>C", "2021_A>T"), N, af.dm, spatial_coords)
}
