% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/LineagePlot.R
\name{PlotVAFHeatmap}
\alias{PlotVAFHeatmap}
\title{Plot Heatmap of VAF for Variants of Interest}
\usage{
PlotVAFHeatmap(
  voi,
  af.dm,
  spatial_coords,
  VAF_cap = 100,
  ngroups = length(voi)
)
}
\arguments{
\item{voi}{Character vector, list of variants of interest.}

\item{af.dm}{Matrix (variants x spot), allele frequency data per variant per spot.}

\item{spatial_coords}{Data frame (spot x 2), spatial coordinates (SPATIAL_1, SPATIAL_2).}

\item{VAF_cap}{Numeric, maximum value to cap the VAF for visualization (default: 100).}

\item{ngroups}{Numeric, number of groups to cut the variant dendrogram (default: number of variants).}
}
\value{
A heatmap showing the VAF for each variant across spots.
}
\description{
This function visualizes the variant allele frequency (VAF) for a list of variants
in a spatial transcriptomics dataset. The VAF values are extracted from \code{af.dm}
and ordered similarly to the BF heatmap example: variants are clustered and grouped,
and spots are iteratively reâ€ordered by descending VAF values within each variant group.
}
\examples{
PlotVAFHeatmap(c("3054_G>C", "2021_A>T"), af.dm, spatial_coords)
}
